designs/data-lake.md

```diff
@@ -0,0 +1,114 @@
+    - Maintains the raw state of the data source
+    - Is appended incrementally and grows over time
+    - Can be any combination of streaming and batch transactions
+- **Silver**: validated, de-duplicated and enriched
+    - Has undergone validation, quality checks and enrichment
+    - Can be trusted and optimised for downstream analytics
+- **Gold**: curated, filtered and aggregated
+    - Optimised for data consumers
+    - Powers analytics, machine learning, and production applications
+
+We should seek to provide a simple implementation of this structure out-of-the-box, whilst allowing organisations to expand and customise this structure to their needs. More information on zone structures is available [here](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/scenarios/cloud-scale-analytics/best-practices/data-lake-zones).
+
+### Integrating with FlowEHR
+
+#### Configuration
+
+To ensure we don't disrupt existing deployments of FlowEHR, I propose we introduce this component as an opt-in feature. We can add a new block to the configuration as follows:
+
+```yaml
+transform:
+    datalake:
+        # additional configuration
+```
+
+If the `data_lakehouse` block is not present, no changes will be made to the existing behaviour of FlowEHR. If defined, users can specify the zones they wish to deploy as follows:
+
+```yaml
+zones:
+- name: Bronze
+  containers:
+    - name: Raw
+- name: Silver
+  containers:
+    - name: Transformed
+- name: Gold
+  containers:
+    - name: Curated
+```
+
+The objects under zones represent the data lakes to deploy (an Azure Data Lake Storage gen2 Filesystem), and the containers represent the storage container(s) within each filesystem.
+
+#### Infrastructure
+
+The new infrastructure to deploy will live in the `infrastructure/transform` Terraform module and will comprise of the following:
+
+- Storage Account with Heirarchical Namespaces enabled (Azure Data Lake Storage Gen2)
+- ADLS Gen2 Filesystem(s) (for each configured zone)
+- Storage Container(s) (a `azurerm_storage_data_lake_gen2_path` for each configured container)
+- ADLS private networking (private endpoint, dns zone and private link)
+- ADLS role assignments for Azure Data Factory and Databricks
+- Databricks ADLS filesystem mounts for each zone
+- Service principal for Databricks identity access to ADLS
+- Databricks secrets for SP credentials
+- ADF linked service for ADLS
+
+As mentioned, if the `datalake` configuration block is not present, none of this infrastructure will be deployed by utilising a dynamic flag.
+
+
+## Considerations
+
+- What do we do with the SQL DB now or longer term? Do we deprecate it in favour of the data lake?
+
+Given that the current "gold" layer in FlowEHR is the SQL DB, we should consider whether it would be superseded by ADLS. This will greatly simplify the architecture of FlowEHR; however the SQL Server currently also serves as the query layer for applications and data scientists / ML models.
```

Comment:
the SQL store - wouldn't we add the config to enable opt-in/out for this too, in the same way as the data lake?

Reply:
Yeah was going to suggest that as Azure SQL is quite expensive

Reply:
Agreed!